src/transformers/retrieval_rag.py

```diff
@@ -0,0 +1,337 @@
+            doc_dict["text"] = [doc[0] for doc in docs]
+            doc_dicts.append(doc_dict)
+        return doc_dicts
+
+    def get_top_docs(self, query_vectors: np.array, n_docs: int = 5):
+        aux_dim = np.zeros(len(query_vectors), dtype="float32").reshape(-1, 1)
+        query_nhsw_vectors = np.hstack((query_vectors, aux_dim))
+        _, docs_ids = self.index.search(query_nhsw_vectors, n_docs)
+        vectors = [[self.index.reconstruct(int(doc_id))[:-1] for doc_id in doc_ids] for doc_ids in docs_ids]
+        ids = [[int(self.index_id_to_db_id[doc_id]) for doc_id in doc_ids] for doc_ids in docs_ids]
+        return torch.tensor(ids), torch.tensor(vectors)
+
+
+class HFIndex(Index):
+    """
+    An and index build for an instance of :class:`~nlp.Datasets`. If ``index_path`` is set to ``None``,
+    we load the pre-computed index available with the :class:`~nlp.arrow_dataset.Dataset`, otherwise, we load the index from the indicated path on disk.
+
+    Args:
+        dataset (:obj:`str`, optional, defaults to ``wiki_dpr``):
+            A datatset identifier of the indexed dataset on HuggingFace AWS bucket (list all available datasets and ids with ``nlp.list_datasets()``).
+        dataset_split (:obj:`str`, optional, defaults to ``train``)
+            Which split of the ``dataset`` to load.
+        index_name (:obj:`str`, optional, defaults to ``train``)
+            The index_name of the index associated with the ``dataset``. The index loaded from ``index_path`` will be saved under this name.
+        index_path (:obj:`str`, optional, defaults to ``None``)
+            The path to the serialized faiss index on disk.
+    """
+
+    def __init__(
+        self,
+        dataset,
+        dataset_split,
+        index_name,
+        index_path,
+    ):
+        super().__init__()
+        self.dataset = dataset
+        self.dataset_split = dataset_split
+        self.index_name = index_name
+        self.index_path = index_path
+        self.index = load_dataset(self.dataset, with_index=False, split=self.dataset_split)
+        self._index_initialize = False
+
+    def is_initialized(self):
+        return self._index_initialize
+
+    def init_index(self):
+        if self.index_path is not None:
+            self.index.load_faiss_index(index_name=self.index_name, file=self.index_path)
+        else:
+            self.index = load_dataset(self.dataset, with_embeddings=True, with_index=True, split=self.dataset_split)
+        self._index_initialize = True
+
+    def get_doc_dicts(self, doc_ids):
+        return [self.index[doc_ids[i].tolist()] for i in range(doc_ids.shape[0])]
+
+    def get_top_docs(self, query_vectors, n_docs=5):
+        _, docs = self.index.get_nearest_examples_batch(self.index_name, query_vectors, n_docs)
+        ids = [[int(i) for i in doc["id"]] for doc in docs]
+        vectors = [doc[self.index_name] for doc in docs]
+        return torch.tensor(ids), torch.tensor(vectors)
```

Comment:
@lhoestq would it make sense to incorporate these utilities somehow in `nlp` directly?

Reply:
`docs` is a list of retrieved docs per query (one dictionary of docs per query)

To make things easier the first thing to do would be to fix the type of the "id" column of wiki_dpr from string to int.

If the dataset format is set to `torch` with columns "ids" and "embeddings", then with this change `docs` would look like
```python
[
  {"ids": Tensor([1, 42, 1337, ...]), "embeddings": Tensor(<2d tensor>)},
  {"ids": Tensor([11, 13, 17, ...]), "embeddings": Tensor(<2d tensor>)},
  ...
]
```

which is already a better format that the current one for `docs`

Then I imagine we could go one step further and make `docs` one dictionary with 2D columns with shape (n_queries, n_docs).
Though actually the `embeddings` column would be 3D because it contains 768-d vectors.

Then we would end up directly with `torch.tensor(ids), torch.tensor(vectors)` as the values of the dict.
What do you think ? This is a breaking change but It would only require changes in this PR and maybe in the docs of DPR